pallets/snarcos/src/lib.rs

```diff
@@ -254,29 +254,53 @@ pub mod pallet {
             proof: Vec<u8>,
             public_input: Vec<u8>,
         ) -> Result<(), (Error<T>, Option<Weight>)> {
+            let data_length_limit = T::MaximumDataLength::get() as usize;
+            let data_length_excess = proof.len().saturating_sub(data_length_limit)
+                + public_input.len().saturating_sub(data_length_limit);
             ensure!(
-                proof.len() <= T::MaximumDataLength::get() as usize
-                    && public_input.len() <= T::MaximumDataLength::get() as usize,
-                (Error::<T>::DataTooLong, None)
+                data_length_excess == 0,
+                (
+                    Error::<T>::DataTooLong,
+                    Some(T::WeightInfo::verify_data_too_long(
+                        data_length_excess as u32
+                    ))
+                )
             );
 
+            let proof_len = proof.len() as u32;
             let proof: S::Proof = CanonicalDeserialize::deserialize(&*proof).map_err(|e| {
                 log::error!("Deserializing proof failed: {:?}", e);
-                (Error::<T>::DeserializingProofFailed, None)
+                (
+                    Error::<T>::DeserializingProofFailed,
+                    Some(T::WeightInfo::verify_data_decoding_fails(proof_len)),
+                )
             })?;
 
             let public_input: Vec<S::CircuitField> =
                 CanonicalDeserialize::deserialize(&*public_input).map_err(|e| {
                     log::error!("Deserializing public input failed: {:?}", e);
-                    (Error::<T>::DeserializingPublicInputFailed, None)
+                    (
+                        Error::<T>::DeserializingPublicInputFailed,
+                        Some(T::WeightInfo::verify_data_decoding_fails(
```

Comment:
In `weights.rs`, this function (`verify_data_decoding_fails`) returns constant cost, regardless of the input.

Reply:
nevertheless, the semantics requires passing data length; in particular, I suppose that on another machine, this parameter might be discovered to have some impact

Reply:
If the costs are constant, this function could be parameter-less but it's probably not in our control?

Reply:
out of control